libsplinter/src/rest_api/auth/roles/store/mod.rs

```diff
@@ -0,0 +1,379 @@
+    /// * no ID or an empty ID was provided
+    /// * no display name or an empty display name was provided
+    /// * empty permissions were provided
+    pub fn build(self) -> Result<Role, InvalidStateError> {
+        if self.permissions.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires at least one permission".into(),
+            ));
+        }
+
+        let id = self
+            .id
+            .ok_or_else(|| InvalidStateError::with_message("A role requires an id field".into()))?;
+        if id.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires a non-empty id field".into(),
+            ));
+        }
+        let display_name = self.display_name.ok_or_else(|| {
+            InvalidStateError::with_message("A role requires a display_name field".into())
+        })?;
+
+        if display_name.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires a non-empty display_name field".into(),
+            ));
+        }
+
+        Ok(Role {
+            id,
+            display_name,
+            permissions: self.permissions,
+        })
+    }
+}
+
+/// Updates an existing role.
+///
+/// This builder only allows the updatable fields to be modified.
+pub struct RoleUpdateBuilder {
+    id: String,
+    display_name: Option<String>,
+    permissions: Vec<String>,
+}
+
+impl RoleUpdateBuilder {
+    /// Updates the display name for the new role.
+    pub fn with_display_name(mut self, display_name: String) -> Self {
+        self.display_name = Some(display_name);
+        self
+    }
+
+    /// Updates the permissions for the new role.
+    pub fn with_permissionss(mut self, permissions: Vec<String>) -> Self {
+        self.permissions = permissions;
+        self
+    }
+
+    /// Builds the updated Role.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`InvalidStateError`] under the following conditions:
+    /// * an empty display name was provided
+    /// * empty permissions were provided
+    pub fn build(self) -> Result<Role, InvalidStateError> {
+        if self.permissions.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires at least one permission".into(),
+            ));
+        }
+
+        let display_name = self.display_name.ok_or_else(|| {
+            InvalidStateError::with_message("A role requires a display_name field".into())
+        })?;
+
+        if display_name.is_empty() {
+            return Err(InvalidStateError::with_message(
+                "A role requires a non-empty display_name field".into(),
+            ));
+        }
+
+        Ok(Role {
+            id: self.id,
+            display_name,
+            permissions: self.permissions,
+        })
+    }
+}
+
+/// An identity that may be assigned roles.
+pub enum Identity {
```

Comment:
This is fine for now, but we may want to consider either removing this or renaming it so it doesn't conflict with the `Identity` struct introduced in https://github.com/Cargill/splinter/pull/1104